### PR TITLE
KAFKA-17056 Convert producer state metadata schemas to use generated protocol

### PR DESCRIPTION
**Pull request by brandboat**
related to https://issues.apache.org/jira/browse/KAFKA-17056,

Convert producer state metadata schemas to use generated protocol.
https://github.com/apache/kafka/blob/33f5995ec379f0d18c6981106838c605ee94be7f/storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java#L94

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java [hunk 1]
```diff
@@ -74,37 +71,10 @@ public class ProducerStateManager {
 
     public static final long LATE_TRANSACTION_BUFFER_MS = 5 * 60 * 1000;
 
-    private static final short PRODUCER_SNAPSHOT_VERSION = 1;
```
Comment (chia7712) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:77]: @brandboat could you add a temporary unit test to verify the compatibility? We can move those struct stuff to testing and then generate some random data for comparison.

File: storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java [hunk 2]
```diff
@@ -646,70 +616,69 @@ public Optional<SnapshotFile> removeAndMarkSnapshotForDeletion(long snapshotOffs
     }
 
     public static List<ProducerStateEntry> readSnapshot(File file) throws IOException {
```
Comment (chia7712) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:618]: It seems we don't have unit test for this method. We should add some tests for `readSnapshot` and `writeSnapshot` before introducing the generated code.

I have filed https://issues.apache.org/jira/browse/KAFKA-17135 for you @brandboat please feel free to un-assign if you have no bandwidth.

**Review by chia7712 — COMMENTED** [review]
@brandboat thanks for this patch

File: storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java (hunk 2 above)
Comment (brandboat) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:618]: Thank you 😃

File: storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java (hunk 1 above)
Comment (brandboat) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:77]: > @brandboat could you add a temporary unit test to verify the compatibility? We can move those struct stuff to testing and then generate some random data for comparison.

Thanks, indeed there should have a compatibility test. One thing need to be mentioned is the unit test for ProducerStateManager is in scala, so I have to convert all old implementation in java code to scala code.

Comment (chia7712) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:77]: Given that is a temporary unit test, please feel free to create a Java UT. We will not merge it

Comment (brandboat) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:77]: > Given that is a temporary unit test, please feel free to create a Java UT. We will not merge it

addressed in a985d20

**Comment by chia7712** [comment]
@brandboat Could you please rebase code?

File: storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java [hunk 3]
```diff
@@ -646,71 +616,70 @@ public Optional<SnapshotFile> removeAndMarkSnapshotForDeletion(long snapshotOffs
...
+        short version;
+        ProducerSnapshot producerSnapshot;
+        try {
+            version = byteBuffer.getShort();
+            producerSnapshot = new ProducerSnapshot(new ByteBufferAccessor(byteBuffer), version);
+        } catch (Exception e) {
             throw new CorruptSnapshotException("Snapshot failed schema validation: " + e.getMessage());
         }
+
+        if (version < ProducerSnapshot.LOWEST_SUPPORTED_VERSION || version > ProducerSnapshot.HIGHEST_SUPPORTED_VERSION)
```
Comment (chia7712) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java]: We should check the version before deserializing

File: storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java [hunk 4]
```diff
@@ -646,71 +616,70 @@ public Optional<SnapshotFile> removeAndMarkSnapshotForDeletion(long snapshotOffs
...
+            ProducerSnapshot.ProducerEntry producerEntry = new ProducerSnapshot.ProducerEntry();
+            producerEntry.setProducerId(producerId);
+            producerEntry.setEpoch(entry.producerEpoch());
+            producerEntry.setLastSequence(entry.lastSeq());
+            producerEntry.setLastOffset(entry.lastDataOffset());
+            producerEntry.setOffsetDelta(entry.lastOffsetDelta());
+            producerEntry.setTimestamp(entry.lastTimestamp());
+            producerEntry.setCoordinatorEpoch(entry.coordinatorEpoch());
+            producerEntry.setCurrentTxnFirstOffset(entry.currentTxnFirstOffset().orElse(-1L));
+            producerEntries.add(producerEntry);
```
Comment (chia7712) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java:676]: Could you please use fluent pattern?
```java
            producerEntries.add(new ProducerSnapshot.ProducerEntry()
                .setProducerId(producerIdEntry.getKey())
                .setEpoch(entry.producerEpoch()));
```

**Review by chia7712 — COMMENTED** [review]
@brandboat thanks for your patch

File: storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java [hunk 5]
```diff
@@ -646,71 +616,70 @@ public Optional<SnapshotFile> removeAndMarkSnapshotForDeletion(long snapshotOffs
...
 
     // visible for testing
     public static void writeSnapshot(File file, Map<Long, ProducerStateEntry> entries, boolean sync) throws IOException {
-        Struct struct = new Struct(PID_SNAPSHOT_MAP_SCHEMA);
-        struct.set(VERSION_FIELD, PRODUCER_SNAPSHOT_VERSION);
-        struct.set(CRC_FIELD, 0L); // we'll fill this after writing the entries
-        Struct[] structEntries = new Struct[entries.size()];
-        int i = 0;
+        ProducerSnapshot producerSnapshot = new ProducerSnapshot();
+        producerSnapshot.setCrc(0L); // we'll fill this after writing the entries
```
Comment (chia7712) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java]: The previous impl is writing data into `ByteBuffer`, so it has to put a temporary data as placeholder. We don't need this workaround now, as we are using our powerful serialization, right?

Comment (brandboat) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java]: Pardon me, did you mean we need to abandon the Crc field or something else ?

Comment (brandboat) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java]: Sorry I look too quick... I believe you mean there is no need to write the 0L here.

Comment (brandboat) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java]: Addressed all comments in the latest commits, thank you !

**Comment by chia7712** [comment]
I will take a look after 3.9.0

**Comment by chia7712** [comment]
@brandboat could you please remove the temporary test since I feel it is ready to be merged

**Review by chia7712 — APPROVED** [review]
LGTM

**Pull request merged**